### DISCUSSION TITLE
Pio ASM cycle counter?

### DISCUSSION THREAD

**Discussion by Manjo80** [micropython/micropython#12607] [RP2040 / Pico]
Hallo, I trying to understand Pio Asm and writing a code to use the hdq protocol on my pi pico. 
I can't find a way how to use the regs as counter, to add 2 everytime to the reg X.

this tis the code I'm trying to write using https://wokwi.com/tools/pioasm

`.program hdq
.wrap_target
    pull block
    mov x, osr
    
    set pindirs, 1 ; set pin as output
    set pins, 1 [31] ; set pin low, wait 190 us
    nop [5]
    set pins, 0 [7] ; set pin high, wait 40 us

    bitloop:
        jmp x-- bit_is_one
        set pins, 1 [17]
        set pins, 0 [19]
        jmp bitloop

    bit_is_one:
        set pins, 1 [7]
        set pins, 0 [29]
        jmp bitloop

    SET PINDIRS, 0 ; Set the pin direction to input

    SET X, 2 ; Initialize the X and Y registers to 0/2
        
    loop:
        WAIT PIN, 0 ; Wait for the pin to change state   
        ;MOV X,   ; add 2 to reg X
        JMP loop ; Jump back to the loop
      
    MOV ISR, X
    push block 
    SET X, 2 ; Reset the X register to 2
        
    JMP loop ; Jump back to the loop and start over`

Greeds

**Comment by mendenm** [comment]
The easy way to do this is to use 
jmp(x_dec, "here1")
label("here")

which subtracts one from x.  You can always complement x before returning it, which makes the subtract  look like an add.
